Implement ItemStack and LivingEntity sensitive method to get the FoodProperties for an Item

canitzp opened the pull request:
This pr is a new version for #7702 and #7543.
Both pr got as reply that an stack transformer is added (#7638), but to this day this hasn't happened.

With this approach the stack transformer wouldn't work, since the new method is not only sensitive to ItemStack, but also to LivingEntity. This can be useful to change the food properties for eating entities, currently wolves and cats.

All calls to Item#getFoodProperties() are changed to use the newer method, which also includes changing FoodData#eat(Item, ItemStack). This was done by adding a new eat method, which also includes the LivingEntity as third parameter, but it can be null. This Nullable was added so that Item#getUseDuration(ItemStack) doesn't need an LivingEntity sensitive method too.

It is important to note here that this pr doesn't change the behavior of Item#isEditable(), since this would involve much more methods to be changed and LexManos explicitly didn't want a item to be both, food and not food (https://github.com/MinecraftForge/MinecraftForge/pull/7543#issuecomment-767257386)

**Edit:** There would be a advantage in changing Item#isEditable() too, because Item#getFoodProperties() is annotated with Nullabe, but the vanilla codes never actually checks for null on the returned value, instead it relies on isEditable() on being false. For Vanilla this works, but when someone overwrites isEditable() to be true, but returns null in getFoodProperties() (which should be possible due to the annotation), the game would crash with an NPE at some point.

canitzp committed:
Implement ItemStack and LivingEntity sensitive method to get the FoodProperties for an Item

pupnewfster commented on patches/minecraft/net/minecraft/world/food/FoodData.java.patch:
```diff
@@ -0,0 +1,20 @@
++      this.eat(p_38713_, p_38714_, null);
++   }
++
++   public void eat(Item item, ItemStack stack, @javax.annotation.Nullable net.minecraft.world.entity.LivingEntity entity)
```
I feel like this overloaded eat method should have the item and stack use the srg names and then just be added to forge's exc file so that they can be properly mapped. That way if more things that use them get added later in the method at some point they don't need to be replaced via patch to reference these names.

pupnewfster commented on patches/minecraft/net/minecraft/world/food/FoodData.java.patch:
```diff
@@ -0,0 +1,20 @@
+       this.f_38697_ = Math.min(this.f_38697_ + (float)p_38708_ * p_38709_ * 2.0F, (float)this.f_38696_);
+    }
+ 
++   @Deprecated
```
This should have a comment/javadoc saying to use the more sensitive version in addition to the deprecated annotation.

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeItem.java:
```diff
@@ -764,4 +765,17 @@ default int getDefaultTooltipHideFlags(@Nonnull ItemStack stack)
+     */
+    default FoodProperties getFoodProperties(ItemStack stack, @Nullable LivingEntity entity)
+    {
+        return self().getFoodProperties();
```
The base food properties method should be deprecated with a comment/javadoc saying to use this more sensitive version.

pupnewfster commented on patches/minecraft/net/minecraft/world/food/FoodData.java.patch:
```diff
@@ -0,0 +1,20 @@
++   }
++
++   public void eat(Item item, ItemStack stack, @javax.annotation.Nullable net.minecraft.world.entity.LivingEntity entity)
++   {
```
Bracket should be on previous line as this is part of a patch.

canitzp committed:
- Overloaded method FoodData#eat(...) now uses srg names for the Item and ItemStack parameter
- FoodData#eat(...) now has a deprecation note on what to use in the future.
- Added Deprecation Annotation and comment to Item#getFoodProperties()
- Added Nullable Annotation to IForgeItem#getFoodProperties, which is a bit controversial. I added it now because the vanilla method also is annotated with Nullable. The problem is that it would crash (NPE) in most situation where Item#isEditable() returns true, but getFoodProperties() returns null.

pupnewfster commented on src/main/resources/forge.exc:
```diff
@@ -51,4 +51,5 @@ net/minecraft/world/level/chunk/PalettedContainer.setBits(IZ)V=|p_63122_,forceBi
-net/minecraft/world/level/storage/loot/LootPool.<init>([Lnet/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer;[Lnet/minecraft/world/level/storage/loot/predicates/LootItemCondition;[Lnet/minecraft/world/level/storage/loot/functions/LootItemFunction;Lnet/minecraft/world/level/storage/loot/providers/number/NumberProvider;Lnet/minecraft/world/level/storage/loot/providers/number/Numbe ...
\ No newline at end of file
+net/minecraft/world/level/storage/loot/LootPool.<init>([Lnet/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer;[Lnet/minecraft/world/level/storage/loot/predicates/LootItemCondition;[Lnet/minecraft/world/level/storage/loot/functions/LootItemFunction;Lnet/minecraft/world/level/storage/loot/providers/number/NumberProvider;Lnet/minecraft/world/level/storage/loot/providers/number/Numbe ...
+net/minecraft/world/food/FoodData.eat(Lnet/minecraft/world/item/Item;Lnet/minecraft/world/item/ItemStack;Lnet/minecraft/world/entity/LivingEntity;)V=|p_38713_,p_38714_,entity
```
We try to keep this file in more or less alphabetical order (and which means `world/food` should be before `world/level` (and before `world/item`

canitzp replied:
Oh ok, I'll change it accordingly

canitzp committed:
forge.exc alphabetical order

pupnewfster commented on src/main/resources/forge.exc:
```diff
@@ -43,6 +43,7 @@ net/minecraft/world/item/ItemStack.onItemUse(Lnet/minecraft/world/item/context/U
 net/minecraft/world/item/ItemStack.onItemUseFirst(Lnet/minecraft/world/item/context/UseOnContext;)Lnet/minecraft/world/InteractionResult;=|p_41662_
 net/minecraft/world/item/MapItem.getCustomMapData(Lnet/minecraft/world/item/ItemStack;Lnet/minecraft/world/level/Level;)Lnet/minecraft/world/level/saveddata/maps/MapItemSavedData;=|p_42910_,p_42911_
 net/minecraft/world/item/alchemy/PotionBrewing$Mix.<init>(Lnet/minecraftforge/registries/ForgeRegistryEntry;Lnet/minecraft/world/item/crafting/Ingredient;Lnet/minecraftforge/registries/ForgeRegistryEntry;)V=|p_43536_,p_43537_,p_43538_
+net/minecraft/world/food/FoodData.eat(Lnet/minecraft/world/item/Item;Lnet/minecraft/world/item/ItemStack;Lnet/minecraft/world/entity/LivingEntity;)V=|p_38713_,p_38714_,entity
```
As I said `f` also comes before `i`

canitzp replied:
oh crap, I should probably learn the alphabet xd

canitzp committed:
forge.exc alphabetical order

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeItem.java:
```diff
@@ -764,4 +765,18 @@ default int getDefaultTooltipHideFlags(@Nonnull ItemStack stack)
+     * @param entity The entity which wants to eat the food. Be aware that this can be null!
+     * @return The current FoodProperties for the item.
+     */
+    @Nullable
```
Are there any restrictions on when this should be nullable? After looking at the PR in more depth it seems there are a few places patched that if a modder returns null from here but marks their item as edible a null pointer will be thrown. It is probably worth specifying in the javadocs for this that it shouldn't return null if the item returns true for is edible. (At least assuming that is actually true)

canitzp replied:
Yeah I wrote about exactly this in the description. With the Nullable annotation it is exactly like Vanilla, since this NPE already is there.

pupnewfster replied:
That is why I believed what I was saying is correct that there are restrictions, so then that definitely should be mentioned here (in the javadocs) to make sure modders know they shouldn't just return null unless their item is marked as never being edible.

canitzp replied:
For me the current question is, should I just remove the Nullable annotation? There is no need for it and the method isn't really Nullable anyways

pupnewfster replied:
I feel like yes as the same as the method it wraps around, it just should have a clarification in the javadocs that it only should be null if isEdible is false.

canitzp committed:
add javadoc for the strange @Nullable behaviour

canitzp committed:
Merge remote-tracking branch 'forge/1.18.x' into 1.18.x

# Conflicts:
#	patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch
#	patches/minecraft/net/minecraft/world/entity/animal/Cat.java.patch
#	patches/minecraft/net/minecraft/world/entity/animal/Wolf.java.patch
#	patches/minecraft/net/minecraft/world/entity/player/Player.java.patch
#	patches/minecraft/net/minecraft/world/item/Item.java.patch

canitzp committed:
Update to 1.18.2

canitzp committed:
Added back player patch

Shadows-of-Fire commented:
Is this superseded in any way by #8405?

canitzp commented:
Well there are certainly similarities. Their system is more complex and offers more features, but even with their events, they only control player hunger. My pr is more focused on the eating for every living entity.

Shadows-of-Fire commented:
Please fix the merge conflicts on this PR so it can be reassigned.

canitzp committed:
Merge remote-tracking branch 'forge/1.18.x' into 1.18.x

# Conflicts:
#	patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch
#	patches/minecraft/net/minecraft/world/entity/player/Player.java.patch

canitzp committed:
Resolve conflicts

canitzp commented:
Sorry for the delay. I haven't noticed the conflicts, but they are resolved now.

pupnewfster reviewed (changes requested):
After a bit of internal discussion it was pointed out that as this is a method adding item stack sensitivity, a bouncer should also be added to `IForgeItemStack` (like a bunch of the other methods in it) that make calling `getFoodProperties` easier as instead of calling `stack.getItem().getFoodProperties(stack, entity)` it will just be `stack.getFoodProperties(entity)`

canitzp committed:
Add bouncer method for IForgeItem#getFoodProperties in IForgeItemStack

canitzp commented:
@pupnewfster You are right, the bouncer is a great addition to use the method more conveniently, therefor it is now added and I also make use of it in one occasion.

pupnewfster commented:
Is there any reason to not make use of it in the other locations? As even if you already have the item reference it would make it a bit more readable as then you just switch that item reference to a stack reference vs adding the stack as a parameter.

canitzp commented:
Well I didn't want to change the base code more than needed and in this one location, it was very straightforward. If you want, than I also can implement these changes

LexManos commented:
Yes it is preferred to move things to calling the bouncer to make the code in the patches simpler. It will be the same number of changes, just changing item.getFoodProperties(stack, entity) to stack.getFoodProperties(entity)

canitzp committed:
Use the stack bouncer method everywhere

canitzp commented:
So the requested changes were done. All callings of getFoodProperties now use the stack bouncer method

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -549,6 +558,18 @@
+    }
+ 
+    private void m_21063_(ItemStack p_21064_, Level p_21065_, LivingEntity p_21066_) {
+-      Item item = p_21064_.m_41720_();
+-      if (item.m_41472_()) {
```
Don't change these two lines as that just makes it a larger patch, only change the call to `getFoodProperties`

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/animal/Wolf.java.patch:
```diff
@@ -9,3 +18,13 @@
+    }
+ 
+    public boolean m_6898_(ItemStack p_30440_) {
+-      Item item = p_30440_.m_41720_();
+-      return item.m_41472_() && item.m_41473_().m_38746_();
```
Don't change the first line or the `item#isEdible` call as that just makes it a larger patch, only change the call to `getFoodProperties`

canitzp committed:
Don't change to much

canitzp commented:
@pupnewfster done
